# Лабораторная работа 4 - Построение двухфакторного эксперимента с использованием квадратичной модели
*Широких Артем - АСОиУб-18-1*

## Задание

1. Составить матрицу планирования ортогонального центральнокомпозиционного плана для двух факторов с использованием дополнительного нулевого фактора ($х_0=1$).
2. Провести эксперимент, повторив 5 раз опыты во всех точках факторного пространства (найти значения функции отклика Y согласно варианту, выданному преподавателем).
3. Проверить однородность дисперсии по критерию Кохрена и, если необходимо, подобрать такое m (m – кратность проведения опытов, не больше 5), чтобы дисперсия была однородной.
4. Найти коэффициенты уравнения регрессии для нормализованной системы координат.
5. С помощью критерия Стьюдента оценить значимость коэффициентов регрессии.
6. Проверить адекватность модели оригиналу с помощью критерия Фишера.

## Решение

### 1. Матрица планирования ОЦКП для двух факторов с использованием дополнительного нулевого фактора ($х_0=1$)

Составляем ОЦКП для двух факторов с использованием дополнительного нулевого фактора ($х_0=1$).

In [130]:
from pprint import pprint

x = [
    [1 for i in range(9)],
    [-1, 1, -1, 1, -1, 1, 0, 0, 0],
    [-1, -1, 1, 1, 0, 0, -1, 1, 0]
]

x.append([x1_i ** 2 for x1_i in x[1]])
x.append([x2_i ** 2 for x2_i in x[2]])
x.append([x1_i * x2_i for x1_i, x2_i in zip(x[1], x[2])])

pprint(x)

[[1, 1, 1, 1, 1, 1, 1, 1, 1],
 [-1, 1, -1, 1, -1, 1, 0, 0, 0],
 [-1, -1, 1, 1, 0, 0, -1, 1, 0],
 [1, 1, 1, 1, 1, 1, 0, 0, 0],
 [1, 1, 1, 1, 0, 0, 1, 1, 0],
 [1, -1, -1, 1, 0, 0, 0, 0, 0]]


Центрируем переменные квадратичных членов для обеспечения ортогональности.

$ \beta=\frac{1}{N}\sum X_{4_i} $

In [131]:
n = len(x[0])
beta = 1 / n * sum(x[4])

pprint(n)
print(beta)

9
0.6666666666666666


Используя найденные значения, составляем новую таблицу для ОЦКП.
Для $x_3$ и $x_4$ рассчитываем значения по формуле $x'_{i_j} = x_{i_j} - \beta$.

In [132]:
e_x = x
e_x[3] = [x3_i - beta for x3_i in x[3]]
e_x[4] = [x4_i - beta for x4_i in x[4]]

pprint(e_x)

[[1, 1, 1, 1, 1, 1, 1, 1, 1],
 [-1, 1, -1, 1, -1, 1, 0, 0, 0],
 [-1, -1, 1, 1, 0, 0, -1, 1, 0],
 [0.33333333333333337,
  0.33333333333333337,
  0.33333333333333337,
  0.33333333333333337,
  0.33333333333333337,
  0.33333333333333337,
  -0.6666666666666666,
  -0.6666666666666666,
  -0.6666666666666666],
 [0.33333333333333337,
  0.33333333333333337,
  0.33333333333333337,
  0.33333333333333337,
  -0.6666666666666666,
  -0.6666666666666666,
  0.33333333333333337,
  0.33333333333333337,
  -0.6666666666666666],
 [1, -1, -1, 1, 0, 0, 0, 0, 0]]


Проверим ортогональность переменных, в случае ортогональности переменные будут удовлетворять данному выражению

$\sum x'_{3_j}*x'_{4_j} = 0$

In [133]:
is_ort = sum(x3_i * x4_i for x3_i, x4_i in zip(e_x[3], e_x[4]))

print(is_ort)

0.0


### 2. Эксперимент во всех точках факторного пространства значение функции отклика Y

Пусть дана выборка результатов эксперимента, представленная показателем Y.

In [134]:
from statistics import mean

y = [
    [3.004, 3.031, 3.035, 3.039, 3.001],
    [5.193, 5.152, 5.177, 5.209, 5.151],
    [3.927, 3.95, 3.936, 3.898, 3.897],
    [7.141, 7.099, 7.111, 7.138, 7.097],
    [4.684, 4.697, 4.688, 4.73, 4.729],
    [9.135, 9.123, 9.166, 9.134, 9.117],
    [6.371, 6.403, 6.343, 6.339, 6.337],
    [14.672, 14.68, 14.695, 14.668, 14.672],
    [5.828, 5.847, 5.842, 5.905, 5.886]
]

pprint(y)

[[3.004, 3.031, 3.035, 3.039, 3.001],
 [5.193, 5.152, 5.177, 5.209, 5.151],
 [3.927, 3.95, 3.936, 3.898, 3.897],
 [7.141, 7.099, 7.111, 7.138, 7.097],
 [4.684, 4.697, 4.688, 4.73, 4.729],
 [9.135, 9.123, 9.166, 9.134, 9.117],
 [6.371, 6.403, 6.343, 6.339, 6.337],
 [14.672, 14.68, 14.695, 14.668, 14.672],
 [5.828, 5.847, 5.842, 5.905, 5.886]]


Посчитаем средние значения функций отклика.

$\overline{y_i} = \frac{\sum y_{i_j}}{v}$

In [135]:
y_mean = [mean(y_i) for y_i in y]

pprint(y_mean)

[3.0220000000000002,
 5.1764,
 3.9216,
 7.1172,
 4.7056000000000004,
 9.135,
 6.3586,
 14.6774,
 5.8616]


Посчитаем дисперсии.

$S^2_i = \frac{\sum (y_{i_j} - \overline{y_i})^2}{v - 1}$

In [136]:
y_disp = [sum([(y_i_j - y_i_mean) ** 2 for y_i_j in y_i]) / (len(y_i) - 1) for y_i, y_i_mean in zip(y, y_mean)]

pprint(y_disp)

[0.0003260000000000039,
 0.0006447999999999915,
 0.0005513000000000032,
 0.0004441999999999938,
 0.0004983000000000066,
 0.00035750000000000164,
 0.0008047999999999925,
 0.00011580000000000238,
 0.0010513000000000024]


### 3. Проверка однородности по критерию Кохрена

Для проверки необходимо посчитать критерий Кохрена получаемый из данных по таблице.

In [137]:
g = max(y_disp) / sum(y_disp)

print(g)

0.21929495202336313


$G \approx 0.219$

$G_{крит} = 0.3584$ - табличное значение критерия Кохрена (при $k = 9, f = 4$)

Так как $G < G_{крит}$, то дисперсия является однородной.

### 4. Коэффициенты уравнения регрессии для нормализованной системы координат

Рассчитаем оценки коэффициентов регрессии.

In [138]:
c = [1/6, 1/6, 1/2, 1/2, 1/4]

a = [0] + [c[i] * sum([x_i_j * y_i_mean for x_i_j, y_i_mean in zip(e_x[i + 1], y_mean)]) for i in range(5)]

print(a)

b_0 = mean(y_mean)

print(b_0)

[0, 1.6298999999999997, 1.859866666666667, -3.452899999999998, 0.1448000000000016, 0.2603]
6.6639333333333335


$\overline{b}_0 \approx 6.664$

Подставим рассчитанные коэффициенты в уравнение регрессии и получим значение $\widetilde{Y_i}$.

In [139]:
e_y_mean = [b_0 + sum([a[j] * e_x[j][i] for j in range(len(a))]) for i in range(len(y))]

print(e_y_mean)

[2.3317666666666685, 5.070966666666667, 5.530900000000003, 9.311300000000001, 3.7865333333333338, 7.046333333333333, 7.154266666666666, 10.873999999999999, 8.86933333333333]


Рассчитаем ошибки дисперсий ошибок наблюдения по формуле $S^2=\frac{\sum (\widetilde{Y_i} - \overline{Y_i})^2}{N-k-1}$.

In [140]:
S = sum([(y_i_mean * e_y_i_mean) ** 2 for y_i_mean, e_y_i_mean in zip(y_mean, e_y_mean)]) / (9 - 5 - 1)

print(S)

13435.570100597675


$S^2 \approx 13435.57$

Рассчитаем оценку $\widehat{a}_0 = \widehat{b}_0 - \beta \sum a_{i + 1}$.

In [141]:
a_0 = b_0 - beta * sum(a)

print(a_0)

6.369288888888887


$\widehat{a}_0 \approx 6.369$

### 5. Оценка значимости коэффициентов регрессии

Рассчитаем дисперсию воспроизводимости

In [142]:
S_pow = [i * S for i in [1/9, 1/6, 1/2, 1/4]]

print(S_pow)

[1492.8411222886305, 2239.2616834329456, 6717.785050298838, 3358.892525149419]


### 6. Проверка адекватности модели с помощью критерия Фишера

Найдём остаточную дисперсию по формуле $S^2_{ост} = \frac{\sum(y_{4_i} - y_{5_i})^2}{N - L}$.

In [143]:
S_last = sum([(y_4_i - y_5_i) ** 2 for y_4_i, y_5_i in zip([y_i[3] for y_i in y], [y_i[4] for y_i in y])]) / (n - len(c))

print(S_last)

0.0017902499999999898


$S^2_{ост} \approx 0.00179$

In [144]:
F = sum([(y_i_mean * e_y_i_mean) ** 2 for y_i_mean, e_y_i_mean in zip(y_mean, e_y_mean)]) / S

print(F)

3.0


$F = 3$

$F_{табл} = 6.39$ при уровне значимости $0.05$

Так как расчётное значение $F = 3$ меньше табличного $F_{табл} = 6.39$ то модель
считается адекватной.